presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,739 @@
+    {
+        checkState(startToken >= memorySequenceId.get(), "Invalid startToken for memory pages");
+        int pagesToRemove = toIntExact(token - startToken);
+        checkArgument(pagesToRemove <= pages.size(), "Invalid token");
+
+        long bytesRemoved = 0;
+        for (int i = 0; i < pagesToRemove; i++) {
+            SerializedPage removedPage = pages.remove();
+            bytesRemoved += removedPage.getRetainedSizeInBytes();
+            memorySequenceId.incrementAndGet();
+        }
+
+        totalBufferedPages.addAndGet(-pagesToRemove);
+        totalBufferedBytes.addAndGet(-bytesRemoved);
+        totalPagesRemaining.addAndGet(-pagesToRemove);
+    }
+
+    private long getPagesSize(Collection<SerializedPage> pages)
+    {
+        return pages.stream().mapToLong(SerializedPage::getRetainedSizeInBytes).sum();
+    }
+
+    private long getPagesRows(Collection<SerializedPage> pages)
+    {
+        return pages.stream().mapToLong(SerializedPage::getPositionCount).sum();
+    }
+
+    private class HandleInfo
+    {
+        private final long bytes;
+        private final int pageCount;
+        private final Range<Long> range;
+        private final ListenableFuture<TempStorageHandle> handleFuture;
+
+        public HandleInfo(Range<Long> range, ListenableFuture<TempStorageHandle> handleFuture, long bytes, int pageCount)
+        {
+            this.range = requireNonNull(range, "range is null");
+            this.handleFuture = requireNonNull(handleFuture, "handleFuture is null");
+            this.bytes = bytes;
+            this.pageCount = pageCount;
+        }
+
+        public long getBytes()
+        {
+            return bytes;
+        }
+
+        public int getPageCount()
+        {
+            return pageCount;
+        }
+
+        public Range<Long> getRange()
+        {
+            return range;
+        }
+
+        public ListenableFuture<TempStorageHandle> getHandleFuture()
+        {
+            return handleFuture;
+        }
+
+        public void removeFile()
+        {
+            executor.execute(() -> {
+                try {
+                    tempStorage.remove(context, handleFuture.get());
+                }
+                catch (Exception e) {
+                    e.printStackTrace();
+                }
+            });
+        }
+    }
+
+    @Immutable
+    private static class PendingRead
+    {
+        private final String taskInstanceId;
+        private final long sequenceId;
+        private final DataSize maxSize;
+        private final SettableFuture<BufferResult> resultFuture = SettableFuture.create();
+
+        private PendingRead(String taskInstanceId, long sequenceId, DataSize maxSize)
+        {
+            this.taskInstanceId = requireNonNull(taskInstanceId, "taskInstanceId is null");
+            this.maxSize = requireNonNull(maxSize, "maxSize is null");
+            this.sequenceId = sequenceId;
```

Comment:
Minor: It seems nice to add a check here as well like we did in other methods to make sure. Even though it is internal class, it is better to give hints for future development.

Reply:
I don't think it'd necessary to have a check the sequenceId there, since we're just saving which sequenceId we'd like to start reading from for the `PendingRead`.